### PR TITLE
Fix for [macOS] some UI elements (e.g. list, tree views) seem to have more space between lines now #1674

### PR DESCRIPTION
**Pull request by kohlschuetter**
cocoa: Enforce native item height for Table, Tree, List if necessary

Newer versions of macOS introduced a new style of table rows with a
padding that is larger than previously seen. This broke item height
computations (#677). An attempt to fix this had the unexpected side
effect that now some UI elements in Eclipse IDE (and other apps with
"-Dorg.eclipse.swt.internal.carbon.smallFonts" enabled),
eclise.platform.ui #1674, and it also broke compatibility with older
versions of macOS that did not have the additional padding.

This change, reverts the previous fix and adds logic to enforce
the native item height by default, but only when smallFonts is not
enabled.

The setting can also be controlled by a new System property,
org.eclipse.swt.internal.cocoa.enforceNativeItemHeightMinimum, that, if
set, overrules any default (smallFonts or not). This allows apps to
bring brack the old macOS behavior even when the smallFonts setting is
undesired.

Lastly, we add support to selectively enable/disable the native item
height minimum for Table, Tree and List on a per-instance basis, so
applications can selectively control the behavior for specific use
cases.

Fixes https://github.com/eclipse-platform/eclipse.platform.ui/issues/1674
Fixes https://github.com/eclipse-platform/eclipse.platform.swt/issues/677

### CODE REVIEW

**Comment by Phillipus** [comment]
@kohlschuetter It's not easy to work with this PR as it is.

1. Please don't revert 0032340b
2. Apply one single commit with your PR change in on top of the most recent `master` branch

**Comment by kohlschuetter** [comment]
@Phillipus OK, I've stashed the reverting changes into one commit, which still applies cleanly to master.

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/graphics/Device.java
Comment (Phillipus) [inline bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/graphics/Device.java:30]: This isn't needed. You can reference `display.smallFonts` directly.

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java
Comment (Phillipus) [inline bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java:136]: This returns 24 whereas the calculated height with font metrics as returned by `getItemHeight` is 25. I'm not sure whether that matters?

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java
Comment (Phillipus) [inline bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java:2528]: Users might want small fonts _and_ native item height. See https://github.com/eclipse-platform/eclipse.platform.ui/issues/1674#issuecomment-1956169651

**Review by Phillipus — CHANGES_REQUESTED** [review]

**Comment by Phillipus** [comment]
I'm not sure whether this is the right way to go. This introduces new API (`setEnforceNativeItemHeightMinimum`) and makes assumptions about the default item height based on whether `smallFonts` is set or not. I think it might be better if @lshanmug and @elsazac commented.

**Comment by Phillipus** [comment]
@kohlschuetter Based on this PR I created a hybrid version at #1057 which uses your idea of having the opt-in option, calculation of max height, and also initializes the height in each control's constructor in case users don't use the `smallFonts` option.

**Comment by kohlschuetter** [comment]
@Phillipus 

The native height with smallFonts enabled is smaller than without (14 instead of 24), which is taken into consideration by my patch. I think your original assumption that a padding of "+8" is always true just doesn't hold for the "smallFonts" case.

The question about how something would look better than the current default is yet another problem (I initially conflate that as well). For that you want the ability to specify a separate padding, probably as leading factor based upon the font size. 
This should be added in a separate patch.

I can remove the `setEnforceNativeItemHeightMinimum` part if this is not desired, it's totally optional.

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java
Comment (kohlschuetter) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java:2528]: see my comment below — the native height is smaller since we use `NSFont.smallSystemFontSize()`, which is taken into account.

**Comment by Phillipus** [comment]
> The native height with smallFonts enabled is smaller than without (14 instead of 24), which is taken into consideration by my patch. 

OK, so `smallFonts` needs to be taken into consideration.

> I think your original assumption that a padding of "+8" is always true just doesn't hold for the "smallFonts" case.

I agree that this padding of 8 has to be removed.

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java
Comment (kohlschuetter) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java:136]: It may return an arbitrary value different from 24 (especially if smallFonts is set or when you're on an older macOS). We only use that value as a minimum. In the end, it seems to work out.

**Comment by Phillipus** [comment]
> Why do we need to call setFont(null) in the constructor? 

(To be clear it's `setItemHeight` in `Table` and `Tree` and `setFont` in `List`.)

This was done when I assumed `smallFonts` wasn't a deciding factor. If it is, then you're right.

**Comment by kohlschuetter** [comment]
@Phillipus Right, I forgot to remove the corresponding code from my version of TableFont.

I've now changed the assignment in the constructor to `setEnforceNativeItemHeightMinimum(boolean)`, which solves this case, and it also makes things a bit more semantically readable / when trying to understand why we're calling such code in the constructor: it's always the same method call `setEnforceNativeItemHeightMinimum` for Table, Tree and List — in List there's no setItemHeight)

**Comment by Phillipus** [comment]
@kohlschuetter 

That seems to gave fixed it, thanks.

Why does `org.eclipse.swt.internal.cocoa.enforceNativeItemHeightMinimum` have true and false values rather then just declaring it or not declaring it like `org.eclipse.swt.internal.carbon.smallFonts`?

**Comment by kohlschuetter** [comment]
@Phillipus 

> Why does org.eclipse.swt.internal.cocoa.enforceNativeItemHeightMinimum have true and false values rather than just declaring it or not declaring it as a flag like org.eclipse.swt.internal.carbon.smallFonts?
>

It is important to check the value so we can set the state regardless of what we determined from `smallFonts`, while omitting the property means "don't override the smallFonts state".

>
> Could it be simplified to:
> ...

No because then you couldn't enforce the minimum with `smallFonts` enabled. It would always be `false`.

**Comment by Phillipus** [comment]
> No because then you couldn't enforce the minimum with smallFonts enabled. It would always be false.

If you mean setting `smallFonts` on and `enforceNativeItemHeightMinimum=true` that evaluates to false.

```java
private boolean initDefaultEnforceNativeItemHeightMinimum() {
    boolean enforce = !smallFonts;
    String v = System.getProperty("org.eclipse.swt.internal.cocoa.enforceNativeItemHeightMinimum");
    if (v != null) {
        enforce = "true".equals(v);
    }
    return enforce;
}
```

**Comment by Phillipus** [comment]
I edited my last comment which was wrong.

What I observe is that if `smallFonts` is set, `enforceNativeItemHeightMinimum` has no effect on the height whether it's true or false.

**Comment by kohlschuetter** [comment]
@Phillipus that‘s because the native height happens to be small enough, as discussed before

**Comment by Phillipus** [comment]
> @Phillipus that‘s because the native height happens to be small enough, as discussed before

Yes, I understand that. But effectively it means that the `enforceNativeItemHeightMinimum` property can be ignored if `smallFonts` is set.

**Comment by kohlschuetter** [comment]
No, because the native row height for small fonts could still be a number that is higher than ascent+descent+1. 

Just because we haven’t seen these yet doesn’t mean they don’t or won’t exist. I wouldn’t be surprised, for example, if turning one of the many Accessibility options results in a few more pixels.

The whole misery with the initial bug came from wrong assumptions around pixel math (the fixed 1 or 8 pixels). 

Lastly, what are we trying to optimize here? This check is called just once …

**Comment by Phillipus** [comment]
Fair enough.

As this PR seems to fulfil all requirements, it should be reviewed by others who are code owners and committers.

**Comment by Phillipus** [comment]
@kohlschuetter 

I don't think you can have the following public platform-dependent API:

```java
public boolean isEnforceNativeItemHeightMinimum() {
}

public void setEnforceNativeItemHeightMinimum(boolean enforce) {
}
```
These have to be private or package private.

**Comment by Phillipus** [comment]
@kohlschuetter Thanks for those changes.

The PR commit seems no longer to be based on the most recent `master`. Could you rebase and force push it?

**Comment by Phillipus** [comment]
@kohlschuetter Thanks for the rebase.

I don't know how things work regarding approval of these types of change, but I think that adding a new system property (`org.eclipse.swt.internal.cocoa.enforceNativeItemHeightMinimum`) will probably require some agreement as to whether that is an OK name for the property and, if approved, might require some documentation somewhere. Also, I think there may be other things regarding the house-keeping of the code (should new methods be placed in alphabetical order in the code, for example) that a code owner or committer might advise on. Just a heads up. ;-)

**Comment by Phillipus** [comment]
`org.eclipse.swt.internal.cocoa.enforceNativeItemHeightMinimum` is rather long. Could it be shorter? `org.eclipse.swt.internal.cocoa.nativeItemHeight`

**Comment by lshanmug** [comment]
@kohlschuetter @Phillipus Thanks for your efforts, I'll review the PR today and get back.

**Review by lshanmug — COMMENTED** [review]
Please see comments below. I've not tested the PR, but I assume setting the font to larger and smaller sizes works fine.

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java
Comment (kohlschuetter) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java:90]: We would lose the ability to configure this setting individually, however unless we want setEnforceNativeItemHeightMinimum `public`, I think we could simplify this indeed.

Comment (Phillipus) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java:90]: Indeed. If `enforceNativeItemHeightMinimum` is removed from Table/Tree/List and referenced like `display.enforceNativeItemHeightMinimum` then the `setEnforceNativeItemHeightMinimum` methods in Table/Tree/List becomes:
 
```java
void setEnforceNativeItemHeightMinimum() {
	setItemHeight(null, null, true); // update height
}
```

And maybe that method can be removed and `setItemHeight` called in the constructor?

Comment (Phillipus) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java:90]: > We would lose the ability to configure this setting individuall

I don't think that's possible unless it becomes API which is not really suitable for platform-dependent.

**Comment by lshanmug** [comment]
> These are the permutations of the two system properties ("Native" is large on later macOS, "Medium" is how it was before, "Small" is small fonts with no padding):
>
>| smallFonts | nativeHeight | Item Height |
|--------|--------|--------|
| not set | not set | Native |
| not set | true | Native |
| not set | false | Medium |
| set | not set | Small |
| set | true | Small |
| set | false | Small | 

I'm wondering if the new system property to control the item height should work independently of the small fonts property.  It could be difficult to understand these combinations from end-user POV. 
Also, shouldn't the default value to use native height should be false, so that applications don't get the new behavior unless explicitly opted-in? Sorry if this was already discussed and I've missed it.

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java
Comment (kohlschuetter) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java]: Yeah, the logic is made such that we don't have to make any changes for Eclipse IDE and any other apps using smallFonts.

There is chance that there are apps that expect the non-small non-native item height. To get that config back, now you can simply specify o.e.s.i.cocoa.useNativeItemHeight=true.

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java
Comment (kohlschuetter) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java]: I had hoped to keep the method public. I don't quite understand the process in the Eclipse project to make something API, but you could of course totally have a public method that is only accessible for cocoa-specific code. Anyhow. I've now simplified the patch.

Comment (Phillipus) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java]: If you develop solely on a Mac and see a public API method you might use it forgetting that applies to Mac only. It will compile and run, except when you try it on Windows or Linux...so all API methods should be cross-platform. I think that's right @lshanmug ?

Comment (kohlschuetter) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java]: Ah, you're right. I forgot that this implementation shares the same package as on Linux/Windows. Nevermind.

**Comment by lshanmug** [comment]
> > Also, shouldn't the default value to use native height should be false, so that applications don't get the new behavior unless explicitly opted-in?
> 
> I wondered about that too. But apps and Eclipse itself already have the new behavior if the `smallFonts` property is not set. Run Eclipse 4.30 (i.e before the contentious padding fix) with the `smallFonts` property removed (or a child Eclipse instance) and we see a mixture of large and small row heights. The mixture of different row heights is because of #677

I agree that was the behavior with 4.30. With 4.31, Eclipse has the native item heights even with smallFonts property set. I think both were not as expected or desired. Is it possible to keep the properties separate so that small fonts controls only font and native height controls only the item height?

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java
Comment (lshanmug) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Table.java]: > If you develop solely on a Mac and see a public API method you might use it forgetting that applies to Mac only. It will compile and run, except when you try it on Windows or Linux...so all API methods should be cross-platform. I think that's right @lshanmug ?

That's right @Phillipus. Otherwise we will have to add empty stubs on other platforms to maintain code portability. So, when we create a new API we try to have implementations for atleast 2 platforms.

**Comment by Phillipus** [comment]
> Is it possible to keep the properties separate so that small fonts controls only font and native height controls only the item height?

They are connected. Using `smallfonts` gives an initial smaller item height when a control is created. So it only makes sense to use the smaller padding. Is that right @kohlschuetter ?

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java
Comment (lshanmug) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java]: Didn't see your comment here before, so added the comment below, please see https://github.com/eclipse-platform/eclipse.platform.swt/pull/1117#issuecomment-2007678290

**Comment by kohlschuetter** [comment]
I've updated the PR to remove the `smallFonts` check, and I clarified the commit message.

**Comment by Phillipus** [comment]
> I've updated the PR to remove the smallFonts check

Will that make any difference?

**Comment by Phillipus** [comment]
I think the solution here proposed by @kohlschuetter is the best compromise. If this PR is merged Eclipse users will get back the old behaviour, because Eclipse ships with the `smallFonts` setting. Apps that don't have the `smallFonts` setting (like our RCP app, Archi) will also keep the native item height (larger) but without the inconsistencies in height due to bug #677

**Comment by lshanmug** [comment]
> @lshanmug I've updated the PR to remove the `smallFonts` check, and I clarified the commit message.
> 
> To be clear, the actual "native height" is different when `smallFonts` is enabled. This is due to using a different default NSFont, but it doesn't look like it makes any difference. So I think it's OK to decouple.
> 
> In the end, it all works out, even if the native height is always enforced.

Trying to understand the change - when smallFonts is enabled, smaller row height is already set and gets set as nativeItemHeight. Also, useNativeItemHeight is always true, so setting it to true/false does not make a difference, hope I understood correctly?

**Comment by kohlschuetter** [comment]
Yes, correct. With smallFonts enabled, nativeHeight is enforced but it doesn’t make a difference.

**Comment by lshanmug** [comment]
The changes look good to me.

Only thing I was wondering is do we still need the new system property as it does not make any difference from Eclipse POV and native height is enabled by default. Will `useNativeItemHeight=false` be useful or required from POV of RCP apps (without smallFonts)? 
A system property is just like an API that needs to be documented and maintained, so we could avoid introducing one if it's not really required by anyone.
@Phillipus @kohlschuetter WDYT?

**Comment by Phillipus** [comment]
@lshanmug Yes, we do need that property. If this PR is merged, we intend to offer the option of native and smaller item heights in our RCP app. The default will be the smaller item height.

See https://github.com/archimatetool/archi/commit/b4db9d0564269cec9d6a10f47490506b53f2adfa

**Comment by Phillipus** [comment]
I'm not sure about this:

```java
private boolean initUseNativeItemHeight() {
	boolean enforce = true;
	String v = System.getProperty("org.eclipse.swt.internal.cocoa.useNativeItemHeight");
	if (v != null) {
		enforce = "true".equals(v);
	}
	return enforce;
}
```

`enforce` defaults to true so setting `useNativeItemHeight` to true has no effect. If the property is set to _any_ value, even empty, then enforce will be false. Effectively you could have a property which is a flag that, if present, does not use native item heights (i.e. don't use native heights).

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java
Comment (laeubi) [inline bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java:2520]:
```suggestion
	String v = System.getProperty("org.eclipse.swt.internal.cocoa.useNativeItemHeight", "whatever should be the default");
```

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java
Comment (laeubi) [inline bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java:2522]:
```suggestion
		enforce = Boolean.parseBoolen(v);
```

**Comment by Phillipus** [comment]
...which reduces to:

```java
    private boolean initUseNativeItemHeight() {
        return Boolean.parseBoolean(System.getProperty("org.eclipse.swt.internal.cocoa.useNativeItemHeight", "true"));
    }
```

and maybe that could be in-lined where it is called?

**Review by lshanmug — APPROVED** [review]
@Phillipus Didn't get a chance to test this myself. Please merge if it looks good based on your testing.

**Comment by Phillipus** [comment]
> @Phillipus Didn't get a chance to test this myself. Please merge if it looks good based on your testing.

@lshanmug  I think it looks OK. Someone else will have to merge, I don't have commit rights.

**Comment by Phillipus** [comment]
The system property `org.eclipse.swt.internal.cocoa.useNativeItemHeight` will need to be documented. What is the process for that?

**Comment by Phillipus** [comment]
> I see 2 of the PR checks failing, any idea what could be the issue?

I don't know how to interpret the failures. Can anyone help?

**Comment by lshanmug** [comment]
> The system property `org.eclipse.swt.internal.cocoa.useNativeItemHeight` will need to be documented. What is the process for that?
> 
> (And, BTW, `org.eclipse.swt.internal.cocoa.useNativeItemHeight=true` is a NOP)

We can add the system property to the N&N pages along with all the details.

**Comment by kohlschuetter** [comment]
@Phillipus 
If the Jenkins build still fails, I'd suggest to go ahead and commit regardless. It may be a build system problem. The build succeeds locally even with the command that fails.

Since the build fails with a SIGKILL (error 137), this could be related to the fallout on macOS 14.4, but I don't know your CI setup.

I have already spent too much time with this bug, so I hope you can figure out the rest on your side.
Best of luck.

**Comment by Phillipus** [comment]
> I have already spent too much time with this bug, so I hope you can figure out the rest on your side.
Best of luck.

And so have I. I am not an Eclipse committer, I cannot say what the problem is.

@lshanmug @merks @iloveeclipse Sorry to ping you, but I seem to be left to pick up the pieces here. This is not my PR, I'm just trying to help. Can someone please sort out this mess?

**Comment by merks** [comment]
FYI, I'm trying to find out what's wrong with the builds.  I kicked one off again but it failed the same way...

**Comment by Phillipus** [comment]
> FYI, I'm trying to find out what's wrong with the builds. I kicked one off again but it failed the same way...

@merks Is it worth me creating a draft copy of this PR rebased onto latest master branch to trigger another build?

**Comment by merks** [comment]
I sort of doubt it. But I did have a case where a new pr with the same commit worked so it doesn’t hurt to try.

**Comment by Phillipus** [comment]
Now that @kohlschuetter is no longer engaged in this PR and the tests are failing I can't see how this PR can progress. For my own part, I have tried to advise and help with this issue but I'm afraid I can no longer usefully participate in this as I have spent weeks of my time on this and have to move on. I'll leave a summary of the situation and a suggestion:

1. The original problem was reported at #677 
2. This was addressed by @elsazac in this [commit](https://github.com/eclipse-platform/eclipse.platform.swt/commit/0032340b9b647db3ef6776d6b2e481397bd74b8d)
3. Mac users complained about the new increased (but consistent) row heights [here](https://github.com/eclipse-platform/eclipse.platform.ui/issues/1674) 
4. @kohlschuetter Proposed this PR to restore the original row heights and a better fix for the original problem
5. Build tests are now failing as a result of this PR
7. This situation needs to be resolved for Eclipse 4.32

I suggest one of the two following options:

1. Somebody continue with this PR and solve the failing tests

or 

2. Revert [this commit ](https://github.com/eclipse-platform/eclipse.platform.swt/commit/0032340b9b647db3ef6776d6b2e481397bd74b8d) and go back to a state where we have the old behavior but inconsistent item heights as originally reported in #677

**Comment by iloveeclipse** [comment]
Are you sure test failures related to this PR? Tests on jenkins are running on Linux, this PR touches Mac code.

**Comment by iloveeclipse** [comment]
Github runner shows two test fails onac, so one should analyse and fix that before merge.

```
Error:  Tests run: 129, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 1.170 s <<< FAILURE! -- in org.eclipse.swt.tests.junit.Test_org_eclipse_swt_widgets_List
Error:  org.eclipse.swt.tests.junit.Test_org_eclipse_swt_widgets_List.test_setFontLorg_eclipse_swt_graphics_Font -- Time elapsed: 0.012 s <<< FAILURE!
java.lang.AssertionError: itemHeight=24, lineHeight=24
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.eclipse.swt.tests.junit.Test_org_eclipse_swt_widgets_List.test_setFontLorg_eclipse_swt_graphics_Font(Test_org_eclipse_swt_widgets_List.java:1631)
```

```
Error:  org.eclipse.swt.tests.junit.Test_org_eclipse_swt_widgets_List.test_getItemHeight -- Time elapsed: 0.009 s <<< FAILURE!
java.lang.AssertionError: :a:
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.eclipse.swt.tests.junit.Test_org_eclipse_swt_widgets_List.test_getItemHeight(Test_org_eclipse_swt_widgets_List.java:504)
```

**Comment by kohlschuetter** [comment]
@iloveeclipse @lshanmug I've managed to unbreak the tests.

The test `Test_org_eclipse_swt_widgets_List.test_setFontLorg_eclipse_swt_graphics_Font` failed because of the native minimum height that is now correctly enforced. As expected, in that configuration the item height for 8 and 12 font size is the same. But the test expected a larger size.

The solution is to further increase the font size in the test (see my second commit)

**Comment by iloveeclipse** [comment]
> managed to unbreak the tests.

Thanks, but one is still failing: https://github.com/eclipse-platform/eclipse.platform.swt/runs/22917018117

```
:a:
java.lang.AssertionError: :a:
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.eclipse.swt.tests.junit.Test_org_eclipse_swt_widgets_List.test_getItemHeight(Test_org_eclipse_swt_widgets_List.java:504)
```

**Comment by kohlschuetter** [comment]
@iloveeclipse Right, I missed that one. It should be fixed now.

Please note that some Jenkins build still fails with SIGKILL ("script returned exit code 137"). I think that's a problem in your build system, unrelated to this change.

**Comment by iloveeclipse** [comment]
> Please note that some Jenkins build still fails with SIGKILL ("script returned exit code 137"). I think that's a problem in your build system, unrelated to this change.

May I ask you to create a trivial dummy change in Mac related SWT code and push a draft PR based on latest master commit to see if that would also fail Linux build? If so, please create new ticket to investigate.

**Comment by merks** [comment]
Note that yesterday, this "test" PR was created and it didn't fail the build...

https://github.com/eclipse-platform/eclipse.platform.swt/pull/1126

So I will push these changes to a new PR and see if that PR works...

**Comment by merks** [comment]
By some magic I don't understand, pushing this makes the builds here start again but they actually link to the build from this PR:

https://github.com/eclipse-platform/eclipse.platform.swt/pull/1128

**Comment by iloveeclipse** [comment]
> Note that yesterday, this "test" PR was created and it didn't fail the build...
> 
> #1126
> 
> So I will push these changes to a new PR and see if that PR works...

My point was that **if** a **dummy** change on Mac SWT code fails Linux build, we should investigate & fix that issue.

**Comment by merks** [comment]
@kohlschuetter 

This will need to be squashed into a single commit.  (It's typically better to amend a commit and force push it to avoid multiple commits in a PR).    But please hold off until the builds are completed so we can verify that all builds are successful before change something again...

**Comment by Phillipus** [comment]
> My point was that **if** a **dummy** change on Mac SWT code fails Linux build, we should investigate & fix that issue.

https://github.com/eclipse-platform/eclipse.platform.swt/pull/1129

**Comment by merks** [comment]
This build fails only for Windows now, but that looks unrelated.  (What a nightmare.)

@kohlschuetter 

Please squash to create a single commit with a single description.  (Thanks for your help in the face of what are an awful lot of frustrations!)

**Comment by iloveeclipse** [comment]
> > My point was that **if** a **dummy** change on Mac SWT code fails Linux build, we should investigate & fix that issue.
> 
> #1129

This one is green.

**Comment by kohlschuetter** [comment]
@merks Done (squashed). Let's see what this run brings. Thanks for your help!

**Comment by merks** [comment]
Yeah, everything was green!  Thanks **everyone** involved!!  🏆 💟

**Comment by lshanmug** [comment]
Thanks everyone for your efforts in resolving the issue!

**Comment by Phillipus** [comment]
> > The system property `org.eclipse.swt.internal.cocoa.useNativeItemHeight` will need to be documented. What is the process for that?
> 
> We can add the system property to the N&N pages along with all the details.

This still needs to be done before 4.32 (I am not volunteering).

**Comment by arauchberger** [comment]
Hi,
i somehow lost overview what i finally should do to get rid of this big spacing since Eclipse 2024-03 (4.31.0).

i have tried all of these (in eclipse.ini) without any effect:

```
-Dorg.eclipse.swt.internal.carbon.smallFonts
-Dorg.eclipse.swt.internal.cocoa.smallItemHeight
-Dorg.eclipse.swt.internal.cocoa.useNativeItemHeight
-Dorg.eclipse.swt.internal.cocoa.itemPadding=1
```

i would be happy if one could clarify what the right approach should be.

thanks in advance

**Comment by kohlschuetter** [comment]
@arauchberger The easiest way would be to update to [Eclipse 2024-06 M1](https://www.eclipse.org/downloads/packages/release/2024-06/m1).

**Pull request merged**
